feat(ivy): ngcc - add ModuleWithProviders processing

petebacondarwin opened the pull request:
Only the last commit is relevant. The rest are inherited from #26906

mary-poppins commented:
You can preview 391883a at https://pr27326-391883a.ngbuilds.io/.

JoostK commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1015,6 +1046,33 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+                prop =>
+                    !!prop.name && ts.isIdentifier(prop.name) && prop.name.text === 'ngModule') ||
+        null;
+    const ngModule = ngModuleProperty && node && ts.isPropertyAssignment(ngModuleProperty) &&
```
The `&& node` check seems superfluous.

petebacondarwin replied:
+1

JoostK commented on packages/compiler-cli/src/ngcc/src/rendering/renderer.ts:
```diff
@@ -160,26 +166,43 @@ export abstract class Renderer {
-      });
-      this.addExports(outputText, dtsFile.fileName, dtsExports);
-    }
+    this.addExports(outputText, dtsFile.fileName, renderInfo.privateExports);
```
👍

JoostK commented on packages/compiler-cli/src/ngcc/src/rendering/renderer.ts:
```diff
@@ -160,26 +166,43 @@ export abstract class Renderer {
+  /**
+   * Add the type parameters to the appropriate functions that return `ModuleWithProviders`
+   * structures.
+   * This function only gets called on typings files, so it doesn't need to be abstract.
```
I think the `so it doesn't need to be abstract` is a bit too abstract ;p I propose `so it doesn't need to be implemented differently for the various bundle formats.`

petebacondarwin replied:
+1

JoostK commented on packages/compiler-cli/src/ngcc/test/host/esm2015_host_spec.ts:
```diff
@@ -485,6 +485,54 @@ const TYPINGS_DTS_FILES = [
+      ngModuleIdentifier() { return { ngModule: InternalModule }; }
+      ngModuleWithEmptyProviders() { return { ngModule: InternalModule, providers: [] }; }
+      ngModuleWithProviders() { return { ngModule: InternalModule, providers: [SomeService] }; }
+      externalNgModule() { return { ngModule: ExternalModule }; }
```
Are these instance methods intentionally named equal to the static counterparts? I don't see why that was chosen.

petebacondarwin replied:
They have the same name because they return the same kind of object.
I guess I should check in the expectations that the methods that we capture are actually the static ones.
I'll make that simple by changing the names of these methods to be prefixed with `instance`.

JoostK commented on packages/compiler-cli/src/ngcc/test/rendering/renderer_spec.ts:
```diff
@@ -278,30 +298,123 @@ A.ngDirectiveDef = ɵngcc0.ɵdefineDirective({ type: A, selectors: [["", "a", ""
+              static withProviders1(): ModuleWithProviders<SomeModule>;
+              static withProviders2(): ModuleWithProviders<SomeModule>;
+              static withProviders3(): ModuleWithProviders<SomeClass>;
+              static withProviders4(): ModuleWithProviders<ɵngcc0.ExternalModule>;`);
```
The expectations for the function declarations are missing.

petebacondarwin replied:
+1

petebacondarwin committed:
feat(ivy): ngcc - map functions as well as classes from source to typings

To support updating `ModuleWithProviders` calls,
we need to be able to map exported functions between
source and typings files, as well as classes.

mary-poppins commented:
You can preview 63e5ba1 at https://pr27326-63e5ba1.ngbuilds.io/.

mary-poppins commented:
You can preview e7119b6 at https://pr27326-e7119b6.ngbuilds.io/.

gkalpak reviewed (approved):
LGTM (I think 😅)

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/analysis/module_with_providers_analyzer.ts:
```diff
@@ -0,0 +1,111 @@
+}
+
+export type ModuleWithProvidersAnalyses = Map<ts.SourceFile, ModuleWithProvidersInfo[]>;
+export const ModuleWithProvidersAnalyses = Map;
```
It makes things work :TM:

gkalpak replied:
A better question would be: Is it used anywhere? :smiley:

petebacondarwin replied:
Good spot. It should be used to create the new instance of the map in `analyzeProgram()`.

petebacondarwin replied:
Fixed

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/rendering/renderer.ts:
```diff
@@ -160,26 +166,43 @@ export abstract class Renderer {
+          info.declaration.type ? info.declaration.type.getStart() : info.declaration.getEnd();
+      const end =
+          info.declaration.type ? info.declaration.type.getEnd() : info.declaration.getEnd();
+      outputText.overwrite(start, end, `ModuleWithProviders<${ngModule}>`);
```
So I don't believe that any generated typings file would ever not have a return type here, so I don't think this case would ever occur. But... you never know someone might have handcrafted a typings file or used a different tool to generate it. So I agree that we should add the `:`.

petebacondarwin replied:
Updated

mary-poppins commented:
You can preview 75d51b4 at https://pr27326-75d51b4.ngbuilds.io/.

mary-poppins commented:
You can preview 0b0eee5 at https://pr27326-0b0eee5.ngbuilds.io/.

alxhub commented on packages/compiler-cli/src/ngcc/src/analysis/module_with_providers_analyzer.ts:
```diff
@@ -0,0 +1,111 @@
+            null;
+        if (!typeParam || isAnyKeyword(typeParam)) {
+          // Either we do not have a parameterized type or the type is `any`.
+          let ngModule = this.host.getDeclarationOfIdentifier(fn.ngModule);
```
Be careful here. This will be a `ts.Declaration` which might be in a completely different source file than the `ModuleWithProviders` function - in which case it will need to be imported into the `.d.ts`. Can you write a test and make sure this case works?

petebacondarwin replied:
Hi Alex - I had tests for NgModules that were imported from a relative file already. But I have now added additional tests for files that are imported from a library too. See https://github.com/angular/angular/pull/27326/commits/c1a3029f5507468d61e6dfd26f84d7b96a71e633
PTAL!

petebacondarwin committed:
feat(ivy): ngcc - add typings to `ModuleWithProviders` functions

Exported functions or static method that return a `ModuleWithProviders`
compatible structure need to provide information about the referenced
`NgModule` type in their return type.

This allows ngtsc to be able to understand the type of `NgModule` that is
being returned from calls to the function, without having to dig into the
internals of the compiled library.

There are two ways to provide this information:

* Add a type parameter to the `ModuleWithProviders` return type. E.g.

```
static forRoot(): ModuleWithProviders<SomeNgModule>;
```

* Convert the return type to a union that includes a literal type. E.g.

```
static forRoot(): (SomeOtherType)&{ngModule:SomeNgModule};
```

This commit updates the rendering of typings files to include this type
information on all matching functions/methods.

mary-poppins commented:
You can preview c1a3029 at https://pr27326-c1a3029.ngbuilds.io/.

JoostK reviewed (commented):
I had a go with this in real life and experienced some problems:

1. Private modules would not be considered by the `PrivateDeclarationsAnalyzer` as the module references were not reported to the `ReferencesRegistry`. In practice, `@ngrx/effects` could not be used because of this problem.
2. Module imports in the typings file would retain their `.d.ts` extension in the import specifier, which is disallowed by TypeScript.

Above two issues are solved in https://github.com/JoostK/angular/commit/fcdbedef0de36d77bf6823ab14a2784b8381323f, feel free to cherry-pick for it to become part of this PR.

3. ngtsc was unable to resolve the qualified name in the form of `ngcc0.SomeModule` that was generated by ngcc. I opened #27562 to add support for that.

petebacondarwin commented:
> 1. Private modules would not be considered by the PrivateDeclarationsAnalyzer as the module references were not reported to the ReferencesRegistry. In practice, @ngrx/effects could not be used because of this problem.

Hi @JoostK I thought about this but wondered if this was a reasonable constraint. But if you are saying that `@ngrx/effects` has this problem then it should be supported. (Event though that is really ugly.)

> 2. Module imports in the typings file would retain their .d.ts extension in the import specifier, which is disallowed by TypeScript.

Good spot.

I'll cherry pick your commit thanks.

mary-poppins commented:
You can preview 4b6c815 at https://pr27326-4b6c815.ngbuilds.io/.

JoostK commented:
> > 1. Private modules would not be considered by the PrivateDeclarationsAnalyzer as the module references were not reported to the ReferencesRegistry. In practice, @ngrx/effects could not be used because of this problem.
> 
> Hi @JoostK I thought about this but wondered if this was a reasonable constraint. But if you are saying that `@ngrx/effects` has this problem then it should be supported. (Event though that is really ugly.)

Yup, see https://github.com/ngrx/platform/blob/9b3c75b620b74ff7c8583dc874d3c00af9f0e3c0/modules/effects/src/effects_module.ts#L15 where they return the internal symbol `EffectsFeatureModule` (well, it is exported privately under a barred-O name).

JoostK commented on packages/compiler-cli/src/ngcc/src/packages/transformer.ts:
```diff
@@ -108,16 +110,26 @@ export class Transformer {
+
+    const moduleWithProvidersAnalyzer =
+        bundle.dts && new ModuleWithProvidersAnalyzer(reflectionHost, referencesRegistry);
+    const moduleWithProvidersAnalyses = moduleWithProvidersAnalyzer &&
```
I think I had a faulty rebase, as the `ModuleWithProviders` analyzer needs to run before the `PrivateDeclarationsAnalyzer` for the registered modules to be actually exported.

petebacondarwin replied:
Thanks

JoostK commented on packages/compiler-cli/src/ngcc/test/rendering/renderer_spec.ts:
```diff
@@ -79,6 +83,12 @@ describe('Renderer', () => {
     contents: `export declare class A {\nfoo(x: number): number;\n}\n`
   };
 
+  const COMPONENT_FILE = {
```
This was also introduced because of a faulty rebase, can be removed . Sorry about that 😞

JoostK reviewed (commented):
I just noticed there's some errors in my commit. Additional fixup commit to address the comments: https://github.com/angular/angular/commit/3870bd3adc884c4edda5f3a1eab6a93aeca33072

petebacondarwin commented:
I took your second commit @JoostK and merged it with your previous one.

mary-poppins commented:
You can preview c5c0d8d at https://pr27326-c5c0d8d.ngbuilds.io/.

mary-poppins commented:
You can preview cd67c12 at https://pr27326-cd67c12.ngbuilds.io/.

alxhub commented on packages/compiler-cli/src/ngcc/src/host/ngcc_host.ts:
```diff
@@ -19,6 +19,21 @@ export function isSwitchableVariableDeclaration(node: ts.Node):
+  /**
+   * The declaration of the function that returns the `ModuleWithProviders` object.
+   */
+  declaration: ts.SignatureDeclaration;
```
Here's a fun one - what happens if this function has multiple call signatures? Each of them would need to be patched. Let's leave this for a future PR though, this is fine as is.

petebacondarwin replied:
Cripes! I'll capture that in an issue

petebacondarwin replied:
https://github.com/angular/angular/issues/27582

alxhub commented on packages/compiler-cli/src/ngcc/src/rendering/renderer.ts:
```diff
@@ -49,6 +50,12 @@ interface DtsClassInfo {
   compilation: CompileResult[];
 }
 
+class DtsRenderInfo {
```
Add a comment describing what this is, and what produces/consumes it.

alxhub commented on packages/compiler-cli/src/ngcc/src/rendering/renderer.ts:
```diff
@@ -160,26 +166,51 @@ export abstract class Renderer {
+        const insertPoint = lastToken && lastToken.kind === ts.SyntaxKind.SemicolonToken ?
+            lastToken.getStart() :
+            info.declaration.getEnd();
+        outputText.appendLeft(insertPoint, `: ModuleWithProviders<${ngModule}>`);
```
👍 please do. There was never any need for user code to use `ModuleWithProviders` at all before.

alxhub commented:
Here's the construction we talked about in the sync meeting:

```typescript
function forRoot(): (Foo|Bar) & {ngModule: RouterModule} {
  return {
    ngModule: RouterModule,
    providers: [...],
    otherType: ...,
  };
}
```

mary-poppins commented:
You can preview f648b5f at https://pr27326-f648b5f.ngbuilds.io/.

mary-poppins commented:
You can preview 1399244 at https://pr27326-1399244.ngbuilds.io/.

petebacondarwin committed:
feat(ivy): support NgModule metadata from calls that do not return ModuleWithProviders types

Normally functions that return `ModuleWithProvider` objects should parameterize
the return type to include the type of `NgModule` that is being returned. For
example `forRoot(): ModuleWithProviders<RouterModule>`.

But in some cases, especially those generated by nccc, these functions to not
explicitly declare `ModuleWithProviders` as their return type. Instead they
return a "intersection" type, one of whose members is a type literal that
declares the `NgModule` type returned. For example:
`forRoot(): CustomType&{ngModule:RouterModule}`.

This commit changes the `NgModuleDecoratorHandler` so that it can extract
the `NgModule` type from either kind of declaration.

mary-poppins commented:
You can preview ef5fff4 at https://pr27326-ef5fff4.ngbuilds.io/.

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -197,9 +197,20 @@ export class NgModuleDecoratorHandler implements DecoratorHandler<NgModuleAnalys
+
+  /**
+   * Retrieve for an `NgModule` identifier (T) within the specified `type`, if it is of the form:
+   * `ModuleWithProvides<T>`
```
```suggestion
   * `ModuleWithProviders<T>`
```

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -231,6 +242,32 @@ export class NgModuleDecoratorHandler implements DecoratorHandler<NgModuleAnalys
+    }
+    for (const t of type.types) {
+      if (ts.isTypeLiteralNode(t)) {
+        for (const m of t.members) {
```
We could constrain the `t.members` list to only contain a single property. No looping should be required here.

petebacondarwin replied:
I was trying to keep it relaxed but perhaps tighter is better. I doubt it really has much of a performance implication.

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -277,3 +314,5 @@ function isDeclarationReference(ref: any): ref is Reference<ts.Declaration> {
        ts.isVariableDeclaration(ref.node));
 }
+
+function hasNgModule
```
Oops :-)

petebacondarwin replied:
gone

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -231,6 +242,32 @@ export class NgModuleDecoratorHandler implements DecoratorHandler<NgModuleAnalys
+    if (!ts.isIntersectionTypeNode(type)) {
+      return null;
+    }
+    for (const t of type.types) {
```
Should we only test the right-most type?

petebacondarwin replied:
Again, I was erring on the side of being more relaxed. We could make it tighter... 🤔

JoostK reviewed (commented):
Nice work! Should we perhaps rename `ngModule` to become ngcc specific, or is this syntax planned to be supported beyond ngcc usecases?

petebacondarwin commented:
I rebased on top of #27562 because otherwise these two PRs would be in conflict. It would be best to land that one first.

mary-poppins commented:
You can preview f00f787 at https://pr27326-f00f787.ngbuilds.io/.

mary-poppins commented:
You can preview 42b916d at https://pr27326-42b916d.ngbuilds.io/.

mary-poppins commented:
You can preview a995ea0 at https://pr27326-a995ea0.ngbuilds.io/.

mary-poppins commented:
You can preview cc79a61 at https://pr27326-cc79a61.ngbuilds.io/.

mary-poppins commented:
You can preview ab1816c at https://pr27326-ab1816c.ngbuilds.io/.

mary-poppins commented:
You can preview 5cd775a at https://pr27326-5cd775a.ngbuilds.io/.

mary-poppins commented:
You can preview c8a98ce at https://pr27326-c8a98ce.ngbuilds.io/.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -183,8 +183,8 @@ export class NgModuleDecoratorHandler implements DecoratorHandler<NgModuleAnalys
-      let typeNode = this.reflector.getDtsDeclarationOfClass(typeRef.node);
-      if (typeNode !== null) {
+      let typeNode = this.reflector.getDtsDeclaration(typeRef.node);
+      if (typeNode !== null && ts.isClassDeclaration(typeNode)) {
```
Should this use `reflector.isClass` instead?

alxhub commented:
http://test/OCL:226226461:BASE:226226508:1545255759970:f6c13c5d

davidjeliba commented:
Hi,

I am receiving the same error as mentioned in https://github.com/angular/angular/issues/29362 but from the @uirouter/angular-hybrid project which also uses ModuleWithProviders.

Has this PR made it to a release? I checked the change log and cannot see a reference.

Thanks,
David.

petebacondarwin commented:
@davidjeliba - yes this PR has been in releases for some months. Can you try testing with 8.1.0-rc.0 and create a new issue if it is still a problem for you?

davidjeliba commented:
Thanks @petebacondarwin, compiling with 8.1.0-rc.0 removed the error. There is a new error however I think that might be an issue with @uirouter/angular-hybrid itself.

Thanks for the quick response, sorry for bothering you on this old PR.

petebacondarwin commented:
Cool. Feel free to create a new issue and mention me in it, so that I will be notified.
